### PR TITLE
fix is_ and is_not

### PR DESCRIPTION
**Pull request by laithsakka**
Fix issue https://github.com/pytorch/pytorch/issues/118805

Note: this was a refresh PR of https://github.com/pytorch/pytorch/pull/118806
discussion there is relevant 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118978

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/builtin.py
Comment (lezcano) [inline torch/_dynamo/variables/builtin.py:1642]: This is expected, as NumPy scalars are treated as 0dim tensors within torch.compile. This is specified in the RFC https://github.com/pytorch/rfcs/pull/54.

File: test/dynamo/test_functions.py
Comment (laithsakka) [reply test/dynamo/test_functions.py:2175]: #  x = np.True_
                    #  z = x | x
                    #  assert(x is z) => should be true, would give false

File: torch/_dynamo/variables/builtin.py
Comment (laithsakka) [reply torch/_dynamo/variables/builtin.py:1642]: are you saying its expected and ok to deviate from python semantics for this case?

Comment (lezcano) [reply torch/_dynamo/variables/builtin.py:1642]: Yes, this is documented. If one wants to fully recover the Python semantics, in most cases it's enough to cast it to the Python type (which will often create a graphbreak, or an unbacked symint).

NumPy maintainers agree that the whole NumPy scalars game should have never happened, but it's too integrated into NumPy to be able to deprecate it.

Comment (laithsakka) [reply torch/_dynamo/variables/builtin.py:1642]: ok would it make sense here then to change the test to actually test reflect the expectations of dynamo , 
maybe also here add something like:

if(testing with dynamo)
   assert (a is b)
if (testing with eager) 
  assert (a is not b)
  
  this way we kind of also establish the expectations of both cases

Comment (lezcano) [reply torch/_dynamo/variables/builtin.py:1642]: Since we don't care about eager... it doesn't matter. You are going to find plenty of tests like this one, that fail for Dynamo because of the different semantics we chose for NumPy scalars. You can either fix them by adding casts to Python scalars, or you can skip them with a comment. I'm fine with either, just choose one (cc @zou3519 in case he has any comments) and let's be consistent.

If I were to choose, I would skip them all with a consistent greppable comment once you find tha the issue is coming from the NumPy scalars vs 0D tensors.

We really don't care about testing eager anymore tho, so don't bother with that part.

Comment (zou3519) [reply torch/_dynamo/variables/builtin.py:1642]: Since we have your blessing to deviate from eager mode, we should probably just skip the eager-mode test and just run these tests with Dynamo

Comment (laithsakka) [reply torch/_dynamo/variables/builtin.py:1642]: thanks guys sounds good will go ahead and update the tests to match dynamo and skip eager

Comment (lezcano) [reply torch/_dynamo/variables/builtin.py]: Eager tests are useful, as they give you a much faster way to run the tests (modulo this small points). Running the whole suite of numpy tests takes less than 2 mins. If you want to deactivate it on CI that's fine with me tho.

File: test/torch_np/numpy_tests/core/test_numeric.py
Comment (lezcano) [inline test/torch_np/numpy_tests/core/test_numeric.py:387]: This is not what I meant. This is not the "desired behaviour", but the behaviour we get from the semantics we chose. What I meant in https://github.com/pytorch/pytorch/pull/118978#discussion_r1476544399 is that this sort of 0D tests can be almost always fixed by doing
```suggestion
        assert_(bool((t | t)) is bool(t))
```

This preserves the semantics of the tests while falling back to eager for the execution of the part that differs between dynamo and eager.

**Review by lezcano — CHANGES_REQUESTED** [review]

File: test/torch_np/numpy_tests/core/test_numeric.py
Comment (anijain2305) [reply test/torch_np/numpy_tests/core/test_numeric.py:387]: Yes, I like @lezcano  suggestion. We can force-cast the outputs so that eager and dynamo behavior are same.

**Comment by laithsakka** [comment]
address comments

**Comment by laithsakka** [comment]
updated as discussed

**Comment by laithsakka** [comment]
@zou3519 @anijain2305 @lezcano  can i have another pass on this. 
re-wrote this as discussed

**Review by lezcano — APPROVED** [review]
Moving forward, we don't really want to add more tests to these files. No need to edit this  PR tho.

**Comment by Skylion007** [comment]
I'm getting this StackTrace on nightly now:
```
[rank6]:[2024-03-06 18:45:34,598] [6/19] torch._dynamo.variables.higher_order_ops: [ERROR] torch._dynamo.exc.Unsupported: comparison PythonModuleVariable() <built-in function is_not> ConstantVariable(NoneType)
```
Is there anyway a None value can even get into a PythonModuleVariable? If not, this seems like we could add a folding constant here.

**Pull request closed without merging**
